api-extractor/src/definitions/ApiItem.ts

```diff
@@ -399,6 +419,150 @@ abstract class ApiItem {
     }
     return false;
   }
+
+  /**
+   * This is called by ApiItems to analyze the types that appear in an expression.  For example,
+   * if a Public API function returns a class that is defined in this package, but not exported,
+   * this is a problem. collectTypeReferences() finds all TypeReference child nodes under the
+   * specified node and analyzes each one.
+   */
+  protected collectTypeReferences(node: ts.Node): void {
+    if (node.kind === ts.SyntaxKind.Block || (node.kind >= 262 && node.kind <= 291)) {
+      // Don't traverse into code blocks or JSDoc items; we only care about the function signature
+      return;
+    }
+
+    if (node.kind === ts.SyntaxKind.TypeReference) {
+      const typeReference: ts.TypeReferenceNode = node as ts.TypeReferenceNode;
+      this._processTypeReference(typeReference);
+    }
+
+    // Recurse the tree
+    for (const childNode of node.getChildren()) {
+      this.collectTypeReferences(childNode);
+    }
+  }
+
+  /**
+   * This is a helper for collectTypeReferences().  It analyzes a single TypeReferenceNode.
+   */
+  private _processTypeReference(typeReferenceNode: ts.TypeReferenceNode): void {
+    const symbol: ts.Symbol = this.extractor.typeChecker.getSymbolAtLocation(typeReferenceNode.typeName);
+    if (!symbol) {
+      // Is this bad?
+      return;
+    }
+
+    if (symbol.flags & ts.SymbolFlags.TypeParameter) {
+      // Don't analyze e.g. "T" in "Set<T>"
+      return;
+    }
+
+    const typeName: string = typeReferenceNode.typeName.getText();
+
+    let currentSymbol: ts.Symbol = symbol;
+    while (true) {
+      const importTypeReference: IImportTypeReference = this._parseImportStatement(currentSymbol);
+      if (importTypeReference) {
+        if (importTypeReference.packageName.substr(0, 2) !== './') {
+          // [CASE 1] External/Resolved
```

Comment:
This algorithm doesn't really work BTW.  The problem with parsing import statements is that the compiler API sometimes skips over them, so you find yourself in CASE 2 trying to code around the problem via heuristics.  The better approach (which I think we chatted about a long time ago and forgot), is like this:

1. Use the while(true) loop to follow the aliases all the way to the ending SourceFile
2. Walk upwards from that directory until you find a directory containing package.json
3. Read the name/version from package.json -- that tells you what package the symbol belongs to.  If it is your own ApiPackage.name/version, then you know it's a local symbol.

As noted in the VSO bug, this has the problem that the declared name may be different from the name that is actually exported by the NPM package (e.g. TraceLogger vs _TraceLogger), but the second half of this bug has a solution for that.